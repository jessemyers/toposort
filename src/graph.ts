export type Node = string;
export type Edge = [Node, Node];

export type NodeSet = Set<Node>;
export type EdgeMap = Map<Node, Set<Node>>;

export interface Graph {
    nodes: NodeSet;
    incoming: EdgeMap;
    outgoing: EdgeMap;
}

export function buildGraph(edges: Edge[]): Graph {
    const nodes = new Set(
        edges.flatMap(
            (edge) => edge,
        ),
    );

    const incoming: EdgeMap = new Map<Node, Set<Node>>();
    const outgoing: EdgeMap = new Map<Node, Set<Node>>();

    for (const node of nodes) {
        if (!incoming.has(node)) {
            incoming.set(node, new Set<Node>());
        }
        if (!outgoing.has(node)) {
            outgoing.set(node, new Set<Node>());
        }
    }

    for (const [fromNode, toNode] of edges) {
        const fromNodes = incoming.get(toNode);
        const toNodes = outgoing.get(fromNode);

        if (fromNodes) {
            fromNodes.add(fromNode);
        }
        if (toNodes) {
            toNodes.add(toNode);
        }
    }

    return {
        nodes,
        incoming,
        outgoing,
    };
}
