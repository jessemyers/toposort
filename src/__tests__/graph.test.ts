import { buildGraph, Edge } from '../graph';

describe('graph', () => {
    it('builds a graph', () => {
        const edges: Edge[] = [
            ['one', 'two'],
            ['two', 'three'],
            ['two', 'four'],
            ['three', 'four'],
        ];

        const graph = buildGraph(edges);
        expect(graph).toMatchSnapshot();
    });
});
