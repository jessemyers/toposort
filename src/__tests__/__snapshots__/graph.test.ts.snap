// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graph builds a graph 1`] = `
Object {
  "incoming": Map {
    "one" => Set {},
    "two" => Set {
      "one",
    },
    "three" => Set {
      "two",
    },
    "four" => Set {
      "two",
      "three",
    },
  },
  "nodes": Set {
    "one",
    "two",
    "three",
    "four",
  },
  "outgoing": Map {
    "one" => Set {
      "two",
    },
    "two" => Set {
      "three",
      "four",
    },
    "three" => Set {
      "four",
    },
    "four" => Set {},
  },
}
`;
